File: client/lib/services/pending_message_queue.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import '../database_service.dart';
import '../models/message_state.dart';
import '../network/socket_client.dart';

class PendingMessageQueue {
  final SocketClient socketClient;
  Timer? _retryTimer;
  bool _isProcessing = false;

  PendingMessageQueue(this.socketClient);

  Future<void> enqueue(Map<String, dynamic> msg) async {
    msg['state'] = MessageState.pending;
    msg['attempt_count'] = 0;
    msg['created_at'] = DateTime.now().toIso8601String();
    
    await DatabaseService.savePendingMessage(msg);
    _scheduleProcess();
  }

  Future<void> onAck(String msgId, int serverTs, int seq) async {
    final pending = DatabaseService.getPendingMessages();
    final msg = pending.firstWhere((m) => m['msg_id'] == msgId, orElse: () => {});
    
    if (msg.isNotEmpty) {
      await DatabaseService.removePendingMessage(msgId);
      
      // Move to main messages box as sent/acked
      await DatabaseService.saveMessage(
        recipientId: msg['recipient_id'],
        senderId: msg['sender_id'],
        text: msg['plaintext'], // Note: this requires plaintext to be in the pending msg
        timestamp: DateTime.fromMillisecondsSinceEpoch(serverTs),
        msgId: msgId,
        status: MessageState.sent,
        isMe: true,
        seq: seq,
      );
    }
  }

  Future<void> onTimeout(String msgId) async {
    final pending = DatabaseService.getPendingMessages();
    final msg = pending.firstWhere((m) => m['msg_id'] == msgId, orElse: () => {});
    
    if (msg.isNotEmpty) {
      int attempts = (msg['attempt_count'] as int?) ?? 0;
      attempts++;
      
      if (attempts >= 5) {
        msg['state'] = MessageState.failed;
        msg['attempt_count'] = attempts;
        await DatabaseService.savePendingMessage(msg);
        
        // Save to main messages as failed so UI shows it
        await DatabaseService.saveMessage(
          recipientId: msg['recipient_id'],
          senderId: msg['sender_id'],
          text: msg['plaintext'],
          timestamp: DateTime.now(),
          msgId: msgId,
          status: MessageState.failed,
          isMe: true,
        );
        await DatabaseService.removePendingMessage(msgId);
      } else {
        msg['state'] = MessageState.retrying;
        msg['attempt_count'] = attempts;
        msg['next_retry_at'] = DateTime.now().add(_getRetryDelay(attempts)).toIso8601String();
        await DatabaseService.savePendingMessage(msg);
        _scheduleProcess();
      }
    }
  }

  void retryAll() {
    final pending = DatabaseService.getPendingMessages();
    for (var msg in pending) {
      if (msg['state'] == MessageState.retrying || msg['state'] == MessageState.pending) {
        // Reset retry time to now to force immediate processing
        msg['next_retry_at'] = DateTime.now().toIso8601String();
        DatabaseService.savePendingMessage(msg);
      }
    }
    _scheduleProcess();
  }

  Duration _getRetryDelay(int attempt) {
    switch (attempt) {
      case 1: return const Duration(seconds: 5);
      case 2: return const Duration(seconds: 10);
      case 3: return const Duration(seconds: 20);
      case 4: return const Duration(seconds: 40);
      default: return const Duration(seconds: 60);
    }
  }

  void _scheduleProcess() {
    if (_isProcessing) return;
    _retryTimer?.cancel();
    _retryTimer = Timer(const Duration(milliseconds: 500), _processQueue);
  }

  Future<void> _processQueue() async {
    if (!socketClient.isConnected) return;
    _isProcessing = true;

    try {
      final pending = DatabaseService.getPendingMessages();
      final now = DateTime.now();

      for (var msg in pending) {
        if (msg['state'] == MessageState.failed) continue;
        
        final nextRetryAtStr = msg['next_retry_at'] as String?;
        final nextRetryAt = nextRetryAtStr != null ? DateTime.parse(nextRetryAtStr) : now;

        if (now.isAfter(nextRetryAt) || now.isAtSameMomentAs(nextRetryAt)) {
          // Process this message
          msg['state'] = MessageState.encrypting;
          await DatabaseService.savePendingMessage(msg);

          try {
            // Note: encryption happens via ChatRepository, so the queue just delegates back or the repository handles encryption before queueing.
            // For architecture purity: The queue should hold encrypted payloads ready to send, OR a callback to encrypt.
            // Since CryptoEngine isolated compute needs plaintext, it's easier if PendingQueue holds plaintext and delegates encryption to ChatRepository.
            // We will expose a stream of "ready to send" messages.
            _readyToSendController.add(msg);
          } catch (e) {
            debugPrint("Encryption error: $e");
            await onTimeout(msg['msg_id']);
          }
        }
      }
    } finally {
      _isProcessing = false;
      // Re-schedule if there are still pending items
      final stillPending = DatabaseService.getPendingMessages().where((m) => m['state'] != MessageState.failed);
      if (stillPending.isNotEmpty) {
        _retryTimer = Timer(const Duration(seconds: 5), _processQueue);
      }
    }
  }

  // Stream for ChatRepository to actually encrypt and send
  final _readyToSendController = StreamController<Map<String, dynamic>>.broadcast();
  Stream<Map<String, dynamic>> get readyToSend => _readyToSendController.stream;

  void dispose() {
    _retryTimer?.cancel();
    _readyToSendController.close();
  }
}

File: android_client/app/src/main/java/com/subhankar/aurachat/service/PendingMessageQueue.kt
package com.subhankar.aurachat.service

import android.util.Log
import com.subhankar.aurachat.data.local.dao.MessageDao
import com.subhankar.aurachat.data.local.dao.PendingMessageDao
import com.subhankar.aurachat.data.local.entity.MessageEntity
import com.subhankar.aurachat.data.local.entity.MessageStatus
import com.subhankar.aurachat.data.local.entity.PendingMessageEntity
import com.subhankar.aurachat.crypto.CryptoEngine
import com.subhankar.aurachat.network.PhoenixSocketManager
import com.subhankar.aurachat.network.AppConfig
import com.subhankar.aurachat.network.ApiClient
import kotlinx.coroutines.*
import java.time.Instant
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.math.pow

/**
 * Pending Message Queue — replaces Dart pending_message_queue.dart.
 *
 * Guarantees message delivery using:
 *   1. Persistent storage (Room) — survives app kill
 *   2. Exponential backoff retries (5s → 10s → 20s → 40s → 60s)
 *   3. Max 5 attempts before marking as FAILED
 *
 * Uses Kotlin Coroutines instead of Dart Timers.
 * Much more powerful: structured concurrency, cancellation, thread safety.
 */
@Singleton
class PendingMessageQueue @Inject constructor(
    private val pendingDao: PendingMessageDao,
    private val messageDao: MessageDao,
    private val cryptoEngine: CryptoEngine,
    private val socketManager: PhoenixSocketManager,
    private val apiClient: ApiClient
) {
    companion object {
        private const val TAG = "PendingQueue"
        private const val MAX_ATTEMPTS = 5
        private const val PROCESS_DELAY_MS = 500L
        private const val RECHECK_INTERVAL_MS = 5_000L
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var processJob: Job? = null

    /**
     * Enqueue a new message for reliable delivery.
     * Replaces: Dart pendingQueue.enqueue(msg)
     */
    suspend fun enqueue(msgId: String, recipientId: String, senderId: String, plaintext: String) {
        val pending = PendingMessageEntity(
            msgId = msgId,
            recipientId = recipientId,
            senderId = senderId,
            plaintext = plaintext,
            state = MessageStatus.PENDING,
            attemptCount = 0,
            createdAt = Instant.now().toString(),
            nextRetryAt = null
        )
        pendingDao.upsert(pending)
        scheduleProcess()
    }

    /**
     * Handle a successful server ACK.
     * Replaces: Dart pendingQueue.onAck(msgId, serverTs, seq)
     */
    suspend fun onAck(msgId: String, serverTs: Long, seq: Int) {
        val pending = pendingDao.getByMsgId(msgId) ?: return

        // Remove from pending queue
        pendingDao.delete(msgId)

        // Update the main message with server-assigned timestamp and seq
        val existing = messageDao.getByMsgId(msgId)
        if (existing != null) {
            messageDao.update(
                existing.copy(
                    status = MessageStatus.SENT,
                    serverTs = serverTs,
                    seq = seq,
                    timestamp = Instant.ofEpochMilli(serverTs).toString()
                )
            )
        }
        Log.d(TAG, "ACK received for $msgId (seq=$seq)")
    }

    /**
     * Handle a send timeout/failure.
     * Replaces: Dart pendingQueue.onTimeout(msgId)
     */
    suspend fun onTimeout(msgId: String) {
        val pending = pendingDao.getByMsgId(msgId) ?: return
        val attempts = pending.attemptCount + 1

        if (attempts >= MAX_ATTEMPTS) {
            // Mark as permanently failed
            pendingDao.upsert(pending.copy(state = MessageStatus.FAILED, attemptCount = attempts))
            messageDao.updateStatus(msgId, MessageStatus.FAILED)
            Log.w(TAG, "Message $msgId FAILED after $MAX_ATTEMPTS attempts")
        } else {
            // Schedule retry with exponential backoff
            val nextRetry = Instant.now().plus(getRetryDelaySeconds(attempts), ChronoUnit.SECONDS)
            pendingDao.upsert(
                pending.copy(
                    state = MessageStatus.RETRYING,
                    attemptCount = attempts,
                    nextRetryAt = nextRetry.toString()
                )
            )
            Log.d(TAG, "Message $msgId retry #$attempts scheduled at $nextRetry")
            scheduleProcess()
        }
    }

    /**
     * Force-retry all pending/retrying messages immediately.
     * Replaces: Dart pendingQueue.retryAll()
     */
    suspend fun retryAll() {
        pendingDao.resetAllRetryTimes(Instant.now().toString())
        scheduleProcess()
    }

    /**
     * Exponential backoff delays — matches your Dart _getRetryDelay().
     */
    private fun getRetryDelaySeconds(attempt: Int): Long = when (attempt) {
        1 -> 5L
        2 -> 10L
        3 -> 20L
        4 -> 40L
        else -> 60L
    }

    // ─── Queue Processing Loop ───────────────────────────────────

    private fun scheduleProcess() {
        processJob?.cancel()
        processJob = scope.launch {
            delay(PROCESS_DELAY_MS)
            processQueue()
        }
    }

    private suspend fun processQueue() {
        if (!socketManager.isConnected) return

        val pending = pendingDao.getRetryableMessages()
        val now = Instant.now()

        for (msg in pending) {
            // Check if it's time to retry
            val nextRetryAt = msg.nextRetryAt?.let { Instant.parse(it) } ?: now
            if (now.isBefore(nextRetryAt)) continue

            // Update state to encrypting
            pendingDao.upsert(msg.copy(state = MessageStatus.ENCRYPTING))

            try {
                // Encrypt and send
                var encryptedPayload: Map<String, Any?>? = null
                
                try {
                    encryptedPayload = cryptoEngine.encryptMessage(msg.recipientId, msg.plaintext)
                } catch (e: Exception) {
                    // No session — fetch bundle and establish
                    Log.d(TAG, "No session for ${msg.recipientId}, fetching keys to establish session...")
                    val bundle = apiClient.fetchPreKeys(msg.recipientId)
                    cryptoEngine.establishSession(msg.recipientId, bundle)
                    
                    // Retry encryption immediately
                    encryptedPayload = cryptoEngine.encryptMessage(msg.recipientId, msg.plaintext)
                }

                val payload = encryptedPayload!!.toMutableMap()
                payload["sender_id"] = msg.senderId

                val response = socketManager.sendMessage(msg.recipientId, msg.msgId, payload)

                // Success! Process ACK
                val serverTs = (response["server_ts"] as? Number)?.toLong() ?: System.currentTimeMillis()
                val seq = (response["seq"] as? Number)?.toInt() ?: 0
                onAck(msg.msgId, serverTs, seq)

            } catch (e: Exception) {
                Log.e(TAG, "Send failed for ${msg.msgId}: ${e.message}")
                onTimeout(msg.msgId)
            }
        }

        // Re-schedule if there are still pending items
        val stillPending = pendingDao.getRetryableMessages()
        if (stillPending.isNotEmpty()) {
            processJob = scope.launch {
                delay(RECHECK_INTERVAL_MS)
                processQueue()
            }
        }
    }

    fun dispose() {
        processJob?.cancel()
        scope.cancel()
    }
}
